### PR TITLE
Ensure that duplicate connections are pruned

### PR DESCRIPTION
**Pull request by uppittu11**
Same as PR #384. 

Makes sure that there is only one connection between any equivalent sets of connection members (e.g. you can't have a bond between atoms 1 and 2 and another one between atoms 2 and 1 since they are the same). This will be useful for applying layered potentials in future PRs.

### CODE REVIEW

**Comment by umesh-timalsina** [comment]
Can you check if the index in 0? This test fails sometimes when the index is 0.

**Comment by uppittu11** [comment]
Nevermind. It looks like it passed in CI

**Review by justinGilmer — APPROVED** [review]
Changes LGTM! One question though, this seems like it will be a quiet thing that occurs, which is fine. But we might want to consider if this should raise a warning or not. If someone tried to add multiple bonds because they were trying to emulate double or triple bonds, it would appear to work, but only ever show one connection

**Review by rsdefever — APPROVED** [review]
Two questions: (1) Why do we `hash(tuple(frozenset(things)))` rather than `hash(frozenset(things))`? 

and (2) right now this evaluates to `False`. Is that the desired/correct behavior?

```python
import gmso
from gmso.core.atom import Atom
from gmso.core.bond import Bond
atom1 = Atom(name="atom1")
atom2 = Atom(name="atom2")

bond1 = Bond(connection_members=[atom1, atom2])
bond2 = Bond(connection_members=[atom2, atom1])
bond1==bond2
```

Other than those questions, the changes look good. Unit tests look thorough. Tests passing locally. LGTM!

File: gmso/core/bond.py [hunk 1]
```diff
@@ -39,6 +39,42 @@ def connection_type(self):
...
+        frozenset
+            A unique set of tuples of equivalent connection members
+        Notes
+        _____
+        For a bond:
+            i, j == j, i
+        where i and j are the connection members.
+        """
+        return frozenset([
+                tuple(self.connection_members),
```
Comment (umesh-timalsina) [inline gmso/core/bond.py:55]: Same comment as above.

**Review by umesh-timalsina — CHANGES_REQUESTED** [review]
As pointed by @rsdefever. The above equality check can be can be achieved by overriding the `__eq__` dunder here. 

```py
class Bond(Connection):
     ....

    def __eq__(self, other):
        if self is other:
            return True
        if self.name == other.name and self._equivalent_member_hash() = other._equivalent_member_hash():
            return True
       else:
            return False
```

**Comment by uppittu11** [comment]
@rsdefever and @umesh-timalsina I think we still want to do ID based hashing/equality for bonds. IIRC we decided that all "physical" objects such as atoms and connections should be hashed via ID. But only when adding to a Topology, where you have a collection of physical objects, you need to make sure you don't have any duplicates. But I'm open to changing this if it is a better option.

**Comment by rsdefever** [comment]
>> IIRC we decided that all "physical" objects such as atoms and connections should be hashed via ID.

IIRC, you are correct. Sorry for trying to throw you off. I didn't realize that this is also `False`:

```python
import gmso
from gmso.core.atom import Atom
from gmso.core.bond import Bond
atom1 = Atom(name="atom1")
atom2 = Atom(name="atom2")

bond1 = Bond(connection_members=[atom1, atom2])
bond2 = Bond(connection_members=[atom1, atom2])
bond1==bond2
```

I still don't understand this:
>> (1) Why do we hash(tuple(frozenset(things))) rather than hash(frozenset(things))?

Apologies if I'm asking an ignorant question.

**Comment by uppittu11** [comment]
@rsdefever Just fixed the `hash(tuple(frozenset(things)))` for the case of bonds and dihedrals. For angles and impropers there are atoms with fixed positions (i.e. not interchangeable with other atoms) so instead of being in a `frozenset` they are stored in a `tuple`, which is ordered, before being hashed.

**Pull request merged**